File: UNXNavigatorDemo/Feature/Common/Controllers/DashboardTabBarController.m
//
//  DashboardTabBarController.m
//  UNXNavigatorDemo
//
//  Created by Leo Lee on 2020/10/8.
//

#import <UNXNavigator/UNXNavigator.h>

#import "DashboardTabBarController.h"
#import "FeatureNavigationController.h"
#import "FeatureTableViewController.h"

#import "UIColor+RandomColor.h"
#import "UIDevice+Additions.h"

@interface DashboardTabBarController () <UITabBarControllerDelegate>

@end

@implementation DashboardTabBarController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    UNXNavigationBarAppearance.standardAppearance.tintColor = [UIColor customTitleColor];
    [UNXNavigationBar registerStandardAppearanceForNavigationControllerClass:[FeatureNavigationController class]];
    
    FeatureTableViewController *featureTableViewController1 = [[FeatureTableViewController alloc] initWithStyle:UITableViewStyleGrouped];
    featureTableViewController1.navigationItem.title = @"UNXNavigationBar 🎉 🎉 🎉";

    UIImage *customNormal = [UIImage imageNamed:@"TabBarCustomNormal"];
    UIImage *customSelected = [UIImage imageNamed:@"TabBarCustomSelected"];
    FeatureNavigationController *navigationController1 = [[FeatureNavigationController alloc] initWithRootViewController:featureTableViewController1];
    navigationController1.tabBarItem = [[UITabBarItem alloc] initWithTitle:@"UNXNavigationBar" image:customNormal selectedImage:customSelected];
    
    FeatureTableViewController *featureTableViewController2 = [[FeatureTableViewController alloc] initWithStyle:UITableViewStyleGrouped];
    featureTableViewController2.navigationItem.title = @"UINavigationBar⚠️⚠️⚠️";

    UIImage *systemNormal = [UIImage imageNamed:@"TabBarSystemNormal"];
    UIImage *systemSelected = [UIImage imageNamed:@"TabBarSystemSelected"];
    BaseNavigationController *navigationController2 = [[BaseNavigationController alloc] initWithRootViewController:featureTableViewController2];
    navigationController2.tabBarItem = [[UITabBarItem alloc] initWithTitle:@"UINavigationBar" image:systemNormal selectedImage:systemSelected];
    
    self.delegate = self;
    self.viewControllers = @[navigationController1, navigationController2];
    
    self.tabBar.tintColor = [UIColor customColorWithLightModeColor:^UIColor * _Nonnull{
        return [UIColor customDarkGrayColor];
    } darkModeColor:^UIColor * _Nonnull{
        return [UIColor customLightGrayColor];
    }];
    
    self.tabBar.unselectedItemTintColor = [UIColor customColorWithLightModeColor:^UIColor * _Nonnull{
        return [UIColor customLightGrayColor];
    } darkModeColor:^UIColor * _Nonnull{
        return [UIColor customDarkGrayColor];
    }];
    
    self.tabBar.translucent = NO; // FIXED: iOS Modal -> Dismiss -> Push, TabBar BUG
    
    // ⚠️Warning!!!
    navigationController2.view.layer.borderWidth = 3.0;
    navigationController2.view.layer.cornerRadius = 40;
    navigationController2.view.layer.borderColor = [UIColor customColorWithLightModeColor:^UIColor * _Nonnull{
        return [UIColor redColor];
    } darkModeColor:^UIColor * _Nonnull{
        return [[UIColor redColor] colorWithAlphaComponent:0.5];
    }].CGColor;
}

- (void)tabBarController:(UITabBarController *)tabBarController didSelectViewController:(UIViewController *)viewController {
    if (UIDevice.isPhoneDevice) return;
    
    NSMutableArray<__kindof UIViewController *> *viewControllers = [NSMutableArray arrayWithArray:self.splitViewController.viewControllers];
    if ([viewController isMemberOfClass:[viewControllers.lastObject class]]) {
        return;
    }
    
    UINavigationController *oldDetailNavigationController = viewControllers.lastObject;
    UIViewController *oldDetailViewController = oldDetailNavigationController.viewControllers.lastObject;
    
    UIViewController *detailViewController = [[[oldDetailViewController class] alloc] init];
    detailViewController.navigationItem.title = oldDetailViewController.navigationItem.title;
    UINavigationController *detailNavigationController = [[[viewController class] alloc] initWithRootViewController:detailViewController];
    
    [viewControllers removeObject:oldDetailNavigationController];
    [viewControllers addObject:detailNavigationController];
    
    [self.splitViewController setViewControllers:viewControllers];
}

@end

File: Example/UNXNavigatorSwiftDemo/Common/Comtrollers/DashboardTabBarController.swift
//
//  DashboardTabBarController.swift
//  UNXNavigatorSwiftDemo
//
//  Created by Leo Lee on 2020/11/23.
//

import UIKit
import UNXNavigator

class DashboardTabBarController: UITabBarController {

    override func viewDidLoad() {
        super.viewDidLoad()
        
        UNXNavigationBar.Appearance.standard.tintColor = .customTitle
        UNXNavigationBar.registerStandardAppearance(forNavigationControllerClass: FeatureNavigationController.self)
        
        let featureTableViewController1 = FeatureTableViewController(style: .grouped)
        let name = (Bundle.main.infoDictionary?[kCFBundleIdentifierKey as String] as? String)?.components(separatedBy: ".").last ?? ""
        featureTableViewController1.navigationItem.title = "\(name) 🎉 🎉 🎉"
        
        let customNormal = UIImage(named: "TabBarCustomNormal")
        let customSelected = UIImage(named: "TabBarCustomSelected")
        let navigationController1 = FeatureNavigationController(rootViewController: featureTableViewController1)
        navigationController1.tabBarItem = UITabBarItem(title: "Custom", image: customNormal, selectedImage: customSelected)
        
        let featureTableViewController2 = FeatureTableViewController(style: .grouped)
        featureTableViewController2.navigationItem.title = "UINavigationBar⚠️⚠️⚠️"
        
        let systemNormal = UIImage(named: "TabBarSystemNormal")
        let systemSelected = UIImage(named: "TabBarSystemSelected")
        let navigationController2 = BaseNavigationController(rootViewController: featureTableViewController2)
        navigationController2.tabBarItem = UITabBarItem(title: "System", image: systemNormal, selectedImage: systemSelected)
        
        delegate = self;
        viewControllers = [navigationController1, navigationController2]

        // ⚠️Warning!!!
        navigationController2.view.layer.borderWidth = 3.0
        navigationController2.view.layer.cornerRadius = 40
        navigationController2.view.layer.borderColor = UIColor.customDynamic(lightMode: { UIColor.red }, darkMode: { UIColor.red.withAlphaComponent(0.5) }).cgColor
        
        tabBar.tintColor = UIColor.customDynamic(lightMode: { .customDarkGray }, darkMode: { () -> UIColor in .customLightGray })
        tabBar.unselectedItemTintColor = UIColor.customDynamic(lightMode: { .customLightGray }, darkMode: { () -> UIColor in .customDarkGray })
        tabBar.isTranslucent = false // FIXED: iOS Modal -> Dismiss -> Push, TabBar BUG
    }

}

extension DashboardTabBarController: UITabBarControllerDelegate {
    func tabBarController(_ tabBarController: UITabBarController, didSelect viewController: UIViewController) {
        if UIDevice.isPhone { return }
        
        guard let last = splitViewController?.viewControllers.last, last.isMember(of: type(of: viewController)) else {
            return
        }
        
        guard var viewControllers = splitViewController?.viewControllers,
              let oldDetailNavigationController = viewControllers.last as? UINavigationController,
              let oldDetailViewController = oldDetailNavigationController.viewControllers.last else { return }
        
        guard let DetailNavigationController = type(of: viewController) as? UINavigationController.Type else { return }
        let detailViewController = type(of: oldDetailViewController).init()
        detailViewController.navigationItem.title = oldDetailViewController.navigationItem.title
        let detailNavigationController = DetailNavigationController.init(rootViewController: detailViewController)

        if let index = viewControllers.firstIndex(of: oldDetailNavigationController) {
            viewControllers.remove(at: index)
            viewControllers.append(detailNavigationController)
            splitViewController?.viewControllers = viewControllers
        }
    }
}
